File: Data Structures & Algorithms/same-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root1
     * @param {TreeNode} root2
     * @return {boolean}
     */
    dfs(root1, root2) {
        if (!root1 && !root2) {
            return true
        }

        if (!root1) {return false}
        if (!root2) {return false}

        if (root1.val !== root2.val) {return false}

        const left = this.dfs(root1.left, root2.left)
        const right = this.dfs(root1.right, root2.right)
        return left && right
    }

    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
    isSameTree(p, q) {
        return this.dfs(p, q)
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree-iii/submission-0.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Parent *Node
 * }
 */

func lowestCommonAncestor(p *Node, q *Node) *Node {
	seen := make(map[*Node]bool)
	for p != nil {
		seen[p] = true
		p = p.Parent
	}
	for q != nil {
		if seen[q] {
			return q
		}
		q = q.Parent
	}
	return nil
}
